httpcore/_trace.py

```diff
@@ -26,6 +27,12 @@ def __init__(
 
     def trace(self, name: str, info: Dict[str, Any]) -> None:
         if self.trace_extension is not None:
+            if inspect.iscoroutinefunction(self.trace_extension):  # pragma: no cover
```

Comment:
It would probably be better to check the return value of the function and see if it's a coroutine  — it's generally more robust to use of decorators and avoids false negatives.

Reply:
Thank you for reviewing, but what if it's not a function at all? Then we can't check the return value because an exception would be thrown; do we need to wrap it in try-except? Isn't it too complicated?

Reply:
If it's not a function then it'll fail with a type error when it's called which seems pretty reasonable. It seems like the goal here is to throw a helpful message and avoid hidden unawaited coroutines. In other cases, like not passing a callable or a callable that takes the wrong arguments, it seems fine to just let the standard errors raise.

Reply:
Yes, it appears I went too far; thanks for the review.🚀

Reply:
👍 thanks for contributing! I think this seems helpful fwiw and checking the return value rather than using inspect is kind of a nit. I've just had difficulties with decorated asynchronous functions in the past and we're less likely to introduce regressions with the alternative approach.